File: services/server/src/safe_socket/safe_socket.py
import socket

def recv_all(socket: socket.socket, size):
    total_bytes_received = 0
    buffer = bytearray()
    while total_bytes_received < size:
        bytes_to_receive = size - len(buffer)
        msg_received = socket.recv(bytes_to_receive)
        if not msg_received:
            return None
        buffer.extend(msg_received)
        total_bytes_received += len(msg_received)
    return bytes(buffer)


def send_all(socket: socket.socket, bytes):
    total_bytes_sent = 0
    bytes_to_send = len(bytes)
    while total_bytes_sent < bytes_to_send:
        bytes_sent = socket.send(bytes[total_bytes_sent:])
        total_bytes_sent += bytes_sent
    return total_bytes_sent

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"io"
)

func SendAll(socket io.Writer, bytes []byte) error {
	total_bytes_sent := 0
	msg_size := len(bytes)
	for total_bytes_sent < msg_size {
		remaining_bytes := bytes[total_bytes_sent:]
		bytes_sent, err := socket.Write(remaining_bytes)
		if err != nil {
			return err
		}
		total_bytes_sent += bytes_sent
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	received_bytes := 0
	buff := make([]byte, size)
	for received_bytes < size {
		req_bytes := buff[received_bytes:]
		bytes_read, err := socket.Read(req_bytes)
		if err != nil {
			return nil, err
		}
		received_bytes += bytes_read
	}
	return buff[:received_bytes], nil
}
